Fix description of empty classes to handle anonymous bit-fields of no…

zygoloid opened the pull request:
…nzero length.

Anonymous bit-fields are not non-static data members, so the definition of "empty class" disregarded them. Clang follows the letter of this rule; GCC and ICC follow the intent, which is that anonymous bit-fields of non-zero length render a class object non-empty.

The C++ standard has the same bug, which is currently in the process of being corrected by http://wg21.link/lwg2358.

rjmccall commented on empty:
```diff
@@ -163,7 +163,8 @@ <h3><a href="#definitions"> 1.1 Definitions </a></h3>
 <dd>
-A class with no non-static data members other than zero-width bitfields,
+A class with no (named) non-static data members,
+no anonymous bit-field members other than zero-width bitfields,
```
Sorry for the very delayed review.  I agree that we should fix the wording here, but I would slightly tweak the proposal:

* In the new clause, "anonymous bit-field members" should just be "unnamed bit-fields" (since "unnamed" is the term from the standard and since unnamed bit-fields are, as noted, not members).

* The "(named)" parenthetical invites confusion because there are other unnamed declarations that do create members (specifically, anonymous structs and unions).  If we feel the need to clarify this, we should add a longer and more explicit note to the new clause.

The standard says that an anonymous bit-field is not a member.  If we ignore that, I think the proposed change is redundant 

I'm not sure I understand how this proposed change actually changes anything.  I understand how the change in the standard's wording was necessary, but this change seems equivalent.  A bit-field is still a non-static data member

 it used to be `\not \exists m \in Members(T) . Variable(m) && !Static(m) &&

zygoloid commented:
I've fixed this paragraph (and the rest of the document) to use the standard spelling "unnamed bit-field", and removed the redundant attempts to clarify with parentheticals.

I also found another two places where we made the same mistake. In particular, this changes the definition of "POD for the purpose of layout" to also exclude classes with overwide unnamed bit-fields (strictly it previously only excluded classes with overwide bit-field members). I believe that change matches the intent, because it's unreasonable to expect an underlying C psABI to specify what to do with overwide bit-fields (which are ill-formed in C).

rjmccall commented:
Looks good, thanks.  I'll leave it open for another week as a prompt for further comment.

rjmccall commented:
Merging.
